src/Servers/IIS/IIS/src/Core/IISHttpContext.cs

```diff
@@ -385,6 +386,15 @@ public unsafe void SetResponseHeaders()
             // This copies data into the underlying buffer
             NativeMethods.HttpSetResponseStatusCode(_pInProcessHandler, (ushort)StatusCode, reasonPhrase);
 
+            if (HttpVersion >= System.Net.HttpVersion.Version20 && NativeMethods.HttpHasResponse4(_pInProcessHandler))
+            {
+                // Check if connection close is set, if so setting goaway
+                if (string.Equals(ConnectionClose, HttpResponseHeaders[HttpKnownHeaderNames.Connection], StringComparison.OrdinalIgnoreCase))
+                {
+                    NativeMethods.HttpSetNeedGoAway(_pInProcessHandler);
```

Comment:
Why is this needed for gRPC? I'm pretty sure Kestrel won't send a GOAWAY if you set a "Connection: close" response header. What happens to concurrent streams?

Reply:
GoAway isn't a gRPC ask, it was an ask from other partners trying to adopt HTTP/2. It was implemented by Http.Sys and IIS at the same time as the other HTTP/2 improvements.

It was first implemented this way in HttpSys and is now being ported to IIS.

Reply:
Who asked? Can you add @halter73, @shirhatti and I to the thread?

Reply:
This should be an explicit feature. If this aborting concurrent streams, that's very different that setting "Connection: close" on an HTTP/1.1 response. If it's draining gracefully, we should probably have a way to specify a timeout.

Reply:
The API was implemented by Windows a while ago, it's too late to add parameters.

This is a graceful GoAway, it does not abort active streams.

The customer explicitly requested the `Connection: close` behavior for consistency with HTTP/1.1.

When we get around to this in Kestrel I would expect there to be a richer feature to let you specify the status and even message, but it might still have a simple fallback for `Connection: close`.

Reply:
Even if it's implemented that way in Http.sys, why wouldn't we do it properly in ASP.NET Core and Connection: close like an implementation detail

Reply:
Let's walk all the way through it after standup.